File: TypeScript/SolidDemo.ts
/* SOLID Principles Demonstration in ONE FILE (TypeScript). */

// =======================
// 1. SINGLE RESPONSIBILITY PRINCIPLE (SRP)
// =======================

// BAD: one class doing multiple things
class UserBad {
  name = '';

  saveToDB(): void {
    console.log('Saving user to DB');
  }

  sendEmail(): void {
    console.log('Sending email');
  }
}

// GOOD: separate responsibilities
class User {
  name = '';
}

class UserRepository {
  save(_user: User): void {
    console.log('Saving user to DB');
  }
}

class EmailService {
  sendEmail(_user: User): void {
    console.log('Sending email');
  }
}

// =======================
// 2. OPEN CLOSED PRINCIPLE (OCP)
// =======================

// BAD: must edit for every new type
class DiscountCalculatorBad {
  calculate(type: string): number {
    if (type === 'NEW') return 10;
    if (type === 'PREMIUM') return 20;
    if (type === 'DIWALI') return 30;
    return 0;
  }
}

// GOOD: extend without modifying
interface DiscountStrategy {
  calculate(): number;
}

class NewCustomerDiscount implements DiscountStrategy {
  calculate(): number {
    return 10;
  }
}

class PremiumCustomerDiscount implements DiscountStrategy {
  calculate(): number {
    return 20;
  }
}

class DiscountCalculator {
  calculate(strategy: DiscountStrategy): number {
    return strategy.calculate();
  }
}

// =======================
// 3. LISKOV SUBSTITUTION PRINCIPLE (LSP)
// =======================

// BAD: violates substitution
class BirdBad {
  fly(): void {
    console.log('Flying');
  }
}

class PenguinBad extends BirdBad {
  fly(): never {
    throw new Error("Can't fly");
  }
}

// GOOD: only flying birds expose fly()
interface Bird {
  eat(): void;
}

interface FlyingBird extends Bird {
  fly(): void;
}

class Sparrow implements FlyingBird {
  eat(): void {
    console.log('Sparrow eating');
  }

  fly(): void {
    console.log('Flying');
  }
}

// =======================
// 4. INTERFACE SEGREGATION PRINCIPLE (ISP)
// =======================

// BAD: fat interface
interface WorkerBad {
  work(): void;
  eat(): void;
}

class RobotBad implements WorkerBad {
  work(): void {
    console.log('Working');
  }

  eat(): never {
    throw new Error("Robot doesn't eat");
  }
}

// GOOD: split interfaces
interface Workable {
  work(): void;
}

interface Eatable {
  eat(): void;
}

class Human implements Workable, Eatable {
  work(): void {
    console.log('Working');
  }

  eat(): void {
    console.log('Eating');
  }
}

class Robot implements Workable {
  work(): void {
    console.log('Working');
  }
}

// =======================
// 5. DEPENDENCY INVERSION PRINCIPLE (DIP)
// =======================

// BAD: high-level depends on low-level
class MySQLDatabase {
  connect(): void {
    console.log('Connecting to MySQL');
  }
}

class ApplicationBad {
  private db = new MySQLDatabase();

  start(): void {
    this.db.connect();
  }
}

// GOOD: depend on an abstraction
interface Database {
  connect(): void;
}

class MySQL implements Database {
  connect(): void {
    console.log('Connecting to MySQL');
  }
}

class PostgreSQL implements Database {
  connect(): void {
    console.log('Connecting to PostgreSQL');
  }
}

class NoSQL implements Database {
  connect(): void {
    console.log('Connecting to NoSQL');
  }
}

class Application {
  constructor(private db: Database) {}

  setDB(db: Database): void {
    this.db = db;
  }

  start(): void {
    this.db.connect();
  }
}

// =======================
// MAIN (TESTING)
// =======================
function main(): void {
  // SRP
  const user = new User();
  new UserRepository().save(user);
  new EmailService().sendEmail(user);

  // OCP
  const calc = new DiscountCalculator();
  console.log(calc.calculate(new PremiumCustomerDiscount()));

  // LSP
  const bird: FlyingBird = new Sparrow();
  bird.fly();

  // ISP
  const robot: Workable = new Robot();
  robot.work();

  // DIP
  const app = new Application(new MySQL());
  app.setDB(new NoSQL());
  app.start();
}

main();

export {};

File: C#/SolidDemo.cs
// SOLID Principles Demonstration in ONE FILE (C#).

using System;

namespace LLDWeekend4.Solid
{
    // =======================
    // 1. SINGLE RESPONSIBILITY PRINCIPLE (SRP)
    // =======================

    // BAD: one class doing multiple things
    public class UserBad
    {
        public string Name = "";
        public void SaveToDb() => Console.WriteLine("Saving user to DB");
        public void SendEmail() => Console.WriteLine("Sending email");
    }

    // GOOD: separate responsibilities
    public class User
    {
        public string Name = "";
    }

    public class UserRepository
    {
        public void Save(User user) => Console.WriteLine("Saving user to DB");
    }

    public class EmailService
    {
        public void SendEmail(User user) => Console.WriteLine("Sending email");
    }

    // =======================
    // 2. OPEN CLOSED PRINCIPLE (OCP)
    // =======================

    // BAD: must modify the class for every new type
    public class DiscountCalculatorBad
    {
        public double Calculate(string type) => type switch
        {
            "NEW" => 10,
            "PREMIUM" => 20,
            "DIWALI" => 30,
            _ => 0,
        };
    }

    // GOOD: extend without modifying
    public interface IDiscountStrategy
    {
        double Calculate();
    }

    public class NewCustomerDiscount : IDiscountStrategy
    {
        public double Calculate() => 10;
    }

    public class PremiumCustomerDiscount : IDiscountStrategy
    {
        public double Calculate() => 20;
    }

    public class DiscountCalculator
    {
        public double Calculate(IDiscountStrategy strategy) => strategy.Calculate();
    }

    // =======================
    // 3. LISKOV SUBSTITUTION PRINCIPLE (LSP)
    // =======================

    // BAD: violates substitution
    public class BirdBad
    {
        public virtual void Fly() => Console.WriteLine("Flying");
    }

    public class PenguinBad : BirdBad
    {
        public override void Fly() => throw new NotSupportedException("Can't fly");
    }

    // GOOD: only flying birds expose Fly()
    public interface IBird
    {
        void Eat();
    }

    public interface IFlyingBird : IBird
    {
        void Fly();
    }

    public class Sparrow : IFlyingBird
    {
        public void Eat() => Console.WriteLine("Sparrow eating");
        public void Fly() => Console.WriteLine("Flying");
    }

    // =======================
    // 4. INTERFACE SEGREGATION PRINCIPLE (ISP)
    // =======================

    // BAD: fat interface
    public interface IWorkerBad
    {
        void Work();
        void Eat();
    }

    public class RobotBad : IWorkerBad
    {
        public void Work() => Console.WriteLine("Working");
        public void Eat() => throw new NotSupportedException("Robot doesn't eat");
    }

    // GOOD: split interfaces
    public interface IWorkable
    {
        void Work();
    }

    public interface IEatable
    {
        void Eat();
    }

    public class Human : IWorkable, IEatable
    {
        public void Work() => Console.WriteLine("Working");
        public void Eat() => Console.WriteLine("Eating");
    }

    public class Robot : IWorkable
    {
        public void Work() => Console.WriteLine("Working");
    }

    // =======================
    // 5. DEPENDENCY INVERSION PRINCIPLE (DIP)
    // =======================

    // BAD: high-level depends on low-level
    public class MySqlDatabase
    {
        public void Connect() => Console.WriteLine("Connecting to MySQL");
    }

    public class ApplicationBad
    {
        private readonly MySqlDatabase _db = new();
        public void Start() => _db.Connect();
    }

    // GOOD: depend on abstraction
    public interface IDatabase
    {
        void Connect();
    }

    public class MySql : IDatabase
    {
        public void Connect() => Console.WriteLine("Connecting to MySQL");
    }

    public class PostgreSql : IDatabase
    {
        public void Connect() => Console.WriteLine("Connecting to PostgreSQL");
    }

    public class NoSql : IDatabase
    {
        public void Connect() => Console.WriteLine("Connecting to NoSQL");
    }

    public class Application
    {
        private IDatabase _db;

        public Application(IDatabase db) => _db = db;

        public void SetDb(IDatabase db) => _db = db;

        public void Start() => _db.Connect();
    }

    // =======================
    // RUNNER
    // =======================
    public static class SolidDemo
    {
        public static void Run()
        {
            // SRP
            var user = new User();
            new UserRepository().Save(user);
            new EmailService().SendEmail(user);

            // OCP
            var calc = new DiscountCalculator();
            Console.WriteLine(calc.Calculate(new PremiumCustomerDiscount()));

            // LSP
            IFlyingBird bird = new Sparrow();
            bird.Fly();

            // ISP
            IWorkable robot = new Robot();
            robot.Work();

            // DIP
            var app = new Application(new MySql());
            app.SetDb(new NoSql());
            app.Start();
        }
    }
}
